# Solutions

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#1.-Files" data-toc-modified-id="1.-Files-1">1. Files</a></span></li></ul></div>

## 1. Files

**1.2.3 File creator function**

Ok, so a function that creates a plaintext file! There are a bunch of ways to do this. But this is how I would do it. As always, I try to base the solutions of what has been taught so far in the course.

First, we import the `Path` class from the `pathlib` module:

In [2]:
from pathlib import Path

Since we will use the `Path` class in our function, we must convert the first parameter to a path object. Why? Because some users may pass a string value as the path, and strings don't have the method `.is_absolute()`. It would crash!

In [35]:
def file_creator(path, content):
    path = Path(path)
    if not path.is_absolute():
        print("The path you provided isn't working.",
              "It should be an absolute path, try again!")
        return
    else:
        if not ".txt" in str(path):
            print("You must include a textfile in your path!")
            return
        else:
            file = open(path,"w")
            file.write(content)

I also included the if-statement `if not ".txt" in str(path):`. If the user forgets to include a filename in their path, the function will crash. The `open()` function needs a filename at the end of the passed path argument. 

The if-statement checks to see "if there isn't a file extension in the path, print a warning!" It does so by checking if the string ".txt" isn't in the path. But this will only work if we convert the path into a string value, hence the `str(path)`. Path objects can't be looped over (they aren't iterable).

If we pass something that isn't an absolute path, such as an empty string, the function will warn us and abort:

In [32]:
file_creator("","Hello")

The path you provided isn't working. It should be an absolute path, try again!


Here's an absolute path to try on:

In [33]:
path = Path.cwd() / "new_file.txt"

In [44]:
file_creator(path,"This is a new file!")

In [45]:
file = open(path,"r")
print(file.read())
file.close()

This is a new file!


In [46]:
# run this if you wan't to delete the file:
path.unlink()

**1.3 Exercise – move the textfiles**

Ok! First we run the provided code to create all the files:

In [63]:
import shutil
import os

from random import randint, seed

seed(30)

# First, create tree of directories
path = Path('exercise')
if os.path.isdir(path):
    shutil.rmtree(path)
    os.mkdir(path)
    os.mkdir(path / 'old_location')
    os.mkdir(path / 'new_location')
else:
    os.mkdir(path)
    os.mkdir(path / 'old_location')
    os.mkdir(path / 'new_location')

# This following code randomly creates 500 files
file_path = Path('exercise/old_location')
for i in range(500):
    # random number to decide file extension of present sequence
    num = randint(0,1)
    # if 'num' equals 0 -> plaintext, otherwise pythonfile
    file_ext = ".txt" if num == 0 else ".py"
    
    # Here to decide file content
    if file_ext == '.txt':
        text = "This is a plaintext file!"
    else:
        text = "# this is a python file"
        
    # finally, writing and closing the file
    file = open(file_path / f"file_{randint(500,10000)}{file_ext}","w")
    file.write(text)
    file.close()


Righty!

So, in this exercise, we need to copy and move a whole bunch of files, using a function, from "old_location" to "new_location". We should also include code that add the current date as part of each files name when moved. Let's do a small TODO schematic:

In [64]:
# TODO – define function

# TODO – convert potential path strings into path objects

# TODO – find all files in old_location, save to list

# TODO – filter all text files

# TODO – loop over our text files

# TODO – rename each file with date

# TODO – create file paths for each file

# TODO – copy each file

# TODO – move each copy to 'new_location'

# TODO – create path variables for the two locations

A pretty scary long list, right? Well, it's actually not that bad, since alot of these TODOs will be done together in one go. Let's start:

In [65]:
# TODO – define function
def move_files(location_A, location_B, date):
    # TODO – convert potential path strings into path objects
    location_A, location_B = Path(location_A), Path(location_B)
    
    # TODO – find all files in old_location, save to list
    all_files = os.listdir(location_A)

    # TODO – filter all text files
    all_txt_files = []
    
    for file in all_files:
        if file.endswith(".txt"):
            all_txt_files.append(file)
        else:
            continue


    # TODO – loop over our text files
    for file_name in all_txt_files:
        
        # TODO – rename each file with date
        new_file_name = f"{date}_{file_name}"
        
        # TODO – create file paths for each file
        file_path_1 = old_loc / file_name
        file_path_2 = new_loc / new_file_name
        
        # TODO – copy each file
        # TODO – move each copy to 'new_location'
        shutil.copy(file_path_1,file_path_2)
    
    print(f"All {len(all_txt_files)} text files copied and moved!")
        
# TODO – create path variables for the two locations
old_loc = Path("exercise/old_location")
new_loc = Path("exercise/new_location")

In [66]:
move_files(old_loc, new_loc, "2020-10-15")

All 262 text files copied and moved!


Let's have a look in the "new_location" to see if it worked (only showing 10 here):

In [67]:
os.listdir(new_loc)[:10]

['2020-10-15_file_4286.txt',
 '2020-10-15_file_7200.txt',
 '2020-10-15_file_3099.txt',
 '2020-10-15_file_2387.txt',
 '2020-10-15_file_2556.txt',
 '2020-10-15_file_6042.txt',
 '2020-10-15_file_8709.txt',
 '2020-10-15_file_6081.txt',
 '2020-10-15_file_1513.txt',
 '2020-10-15_file_6254.txt']

Dates are now added to the files' name! All in all there are 262 text files, which means there are more text files than python files. 

Remember, if you just change the file extension in our function. This would work for any type of file you have, anywhere on your computer! Maybe you can find som use out of it somewhere? Have fun!

(run this following code if you want to remove all exercise folders and files:)

In [68]:
shutil.rmtree("exercise/")